tsdate/approx.py

```diff
@@ -0,0 +1,325 @@
+"""
+Tools for approximating combinations of Gamma variates with Gamma distributions
+"""
+import mpmath
+import numpy as np
+import scipy.stats
+
+
+class GammaDiff:
+    r"""
+    The distribution of the difference of two gamma RVs restricted to the positive reals.
+
+    Specifically, for p(x; a, b) = b^a/Gamma(a) x^{a-1} e^{-xb}, this is
+
+        p(u; a1, a2, b1, b2) \propto \int_u^f p(x; a1, b1) p(x - u; a2, b2) dx, u \geq 0
+    """
+
+    @staticmethod
+    def _2F1(a, b, c, z):
+        """
+        log of |Re| for Gaussian hypergeometric function
+        """
+        val = mpmath.log(mpmath.hyp2f1(a, b, c, z))
+        return float(val)
+
+    @staticmethod
+    def _W(k, u, z):
+        """
+        log of |Re| for Whittaker function of the second kind
+        """
+        val = mpmath.log(mpmath.whitw(k, u, z))
+        return float(val)
+
+    @staticmethod
+    def _Gamma(x):
+        """
+        log of |Re| for Gamma function
+        """
+        val = mpmath.loggamma(x)
+        return float(val)
+
+    @staticmethod
+    def _2F1CF(a, b, c, z, tol=1e-8):
+        """
+        Calculate the quotient, :math:`2F1(a, b; c; z) / 2F1(a+1, b+1; c+1; z)`
+        using continued fractions, where 2F1 is the Gaussian hypergeometric
+        function.  See 15.3.13 of Cuyt et al (2008) "Handbook of Continued Fractions
+        for Special Functions". Also see 15.7.5 in DLMF.
+        """
+        assert z < 0.5, "Continued fraction only convergent for z < 0.5"
+        assert c > 0, "Lower argument of hypergeometric function must be positive"
+        n = 0
+        while True:
+            pn = (a + n) * (b + n) * z * (1 - z)
+            qn = c + n - (a + b + 2 * n + 1) * z
+            if n == 0:
+                P = qn
+                Q = 1
+                P_last = 1
+                Q_last = 0
+                PQ_last = P / Q
+            else:
+                P_tmp = P
+                Q_tmp = Q
+                P = qn * P + pn * P_last
+                Q = qn * Q + pn * Q_last
+                P_last = P_tmp
+                Q_last = Q_tmp
+                PQ = P / Q
+                if abs(PQ - PQ_last) / max(abs(PQ), abs(PQ_last)) < tol:
+                    break
+                PQ_last = PQ
+            n += 1
+        return PQ / c
+
+    def __init__(self, shape1, shape2, rate1, rate2):
+        assert shape1 > 0 and shape2 > 0, "Shape parameters must be positive"
+        assert rate1 > 0 and rate2 > 0, "Rate parameters must be positive"
+        self.shape1 = shape1
+        self.shape2 = shape2
+        self.rate1 = rate1
+        self.rate2 = rate2
+        self.beta = rate1 + rate2
+        self.mu = (1.0 - shape1 - shape2) / 2
+        self.kappa = (shape1 - shape2) / 2
+        self.nu = (shape1 + shape2) / 2
+        self.zeta = 1 / 2 - rate2 / (rate1 + rate2)
+        assert self.zeta > -1 / 2, "Invalid parameters"
+        assert self.nu + 1 / 2 > self.mu, "Invalid parameters"
+
+    def simulate(self, n):
```

Comment:
Should this go in a separate test suite? If not, perhaps a comment to say that this is for testing only, I presume?